libs/grok/src/main/java/org/elasticsearch/grok/GrokCaptureExtracter.java

```diff
@@ -26,14 +27,16 @@ public abstract class GrokCaptureExtracter {
      * is mutable and should be discarded after collecting a single result.
      */
     static class MapExtracter extends GrokCaptureExtracter {
-        private final Map<String, Object> result;
+        private final Map<String, List<Object>> result;
```

Comment:
I've decided to use this type of `Map` since otherwise, the code becomes a bit messy while extracting the elements

Reply:
I see, do you think we could implement this maybe earlier in the process? In the emission of the values maybe? This is just an idea to discuss where it's better to implement this.

I am trying to consider also the performance a bit, adding it here adds a small penalty to every grok processor. I was hoping if we did it closer to the emission we could check if there are multiple entries and only use a list there.

Let me know what you think. I haven't worked  on the solution by the way, I am just sharing my thoughts on the subject and I think it might be worth a small effort to see if it's possible.

Reply:
This is the earlier stage where this can happen because within the `Grok` class we never traverse the list of (potential) captured groups. Thinking about the performance, I had an idea to avoid having to reserve memory in case a group is only being used once (-> do not have to create an ArrayList with 1 element). I'll the update the PR tomorrow